src/ProxyClient/Varnish.php

```diff
@@ -133,26 +142,28 @@ public function refresh($url, array $headers = array())
      *                              refresh and no base URL is set
      *
      */
-    protected function createRequest($method, $url, array $headers = array())
+    protected function queueRequest($method, $url, array $headers = [])
     {
-        $request = parent::createRequest($method, $url, $headers);
+        $request = new InvalidationRequest($method, $url, $headers);
 
-        // For purge and refresh, add a host header to the request if it hasn't
-        // been set
         if (self::HTTP_METHOD_BAN !== $method
-            && '' == $request->getHeader('Host')
+            && !$this->baseUriSet
+            && !$request->getHeaderLine('Host')
         ) {
             throw MissingHostException::missingHost($url);
         }
 
-        return $request;
+        parent::queueRequest($method, $url, $headers);
     }
 
     /**
      * {@inheritdoc}
+     *
+     * Additionally, store whether a base URI was set
      */
-    protected function getAllowedSchemes()
+    protected function initQueue(array $servers, $baseUri)
     {
-        return array('http');
+        $this->queue = new RequestQueue($servers, $baseUri);
+        $this->baseUriSet = $baseUri !== null;
```

Comment:
I’m not too happy with this myself, but we need it for Varnish to [throw the `MissingHostException()`](https://github.com/FriendsOfSymfony/FOSHttpCache/pull/173/files#diff-006816ea733cc0564d6a5b345ed47929R153). That exception is thrown when you try to invalidate a relative URL but have no base URI set. This makes no sense in Varnish, because the hash always includes the full URL, so you’ll be purging something non-existent. Does this work the same for NGINX and Symfony HttpCache? If so, the exception could be moved to `RequestQueue`. Except perhaps for the BAN method check, because BANs don’t need a full URL.